src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,183 @@
+    private Node<T> last;
+
+    MyLinkedList() {
+    }
+
     @Override
     public void add(T value) {
+        if (size == 0) {
+            first = new Node<>(null, value, null);
+            last = first;
+        } else {
+            Node<T> node = new Node<>(last, value, null);
+            last.next = node;
+            last = node;
+        }
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
+        validateAddIndex(index);
+        Node<T> newNode = new Node<>(null, value, null);
+        if (size == 0) {
+            first = newNode;
+            last = newNode;
+        } else if (index == 0) {
+            newNode.next = first;
+            first.prev = newNode;
+            first = newNode;
+        } else if (index == size) {
+            last.next = newNode;
+            newNode.prev = last;
+            last = newNode;
+        } else {
+            Node<T> node = getNodeByIndex(index);
+            newNode.prev = node.prev;
+            node.prev.next = newNode;
+            node.prev = newNode;
+            newNode.next = node;
+        }
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (int i = 0; i < list.size(); i++) {
+            add(list.get(i));
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        validateIndex(index);
+        Node<T> node = getNodeByIndex(index);
+        return node.item;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        validateIndex(index);
+        Node<T> node = getNodeByIndex(index);
+        if (node != null) {
+            final T replacedElement = node.item;
+            node.item = value;
+            return replacedElement;
+        } else {
+            return null;
+        }
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        validateIndex(index);
+        Node<T> node = getNodeByIndex(index);
+        final T removedElement = node.item;
+        unlink(node);
+        return removedElement;
     }
 
     @Override
     public boolean remove(T object) {
-        return false;
+        Node<T> node = getNodeByValue(object);
+        return unlink(node);
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private void validateIndex(int index) {
+        if (index >= size || index < 0) {
+            throw new IndexOutOfBoundsException("Incorrect index provided");
+        }
+    }
+
+    private void validateAddIndex(int index) {
+        if (index > size || index < 0) {
+            throw new IndexOutOfBoundsException("Incorrect index provided");
+        }
+    }
+
+    private Node<T> getNodeByIndex(int index) {
+        Node<T> node;
+        int middle = size / 2;
+        if (index <= middle) {
+            node = first;
```

Comment:
Extract iteration from head and tail to separate methods

Reply:
done